From 16388b8027774b3bf1cdfd5b1582231befdda05c Mon Sep 17 00:00:00 2001
From: "Nguyen Van Tu [FGA.AIS]" <TuNV16@fsoft.com.vn>
Date: Wed, 14 Nov 2018 16:22:29 +0700
Subject: [PATCH 402/628] drm: adv7511: Fix get-mode error when changing LVDS
 to HDMI display

- Get-mode error occurs in this condition:
    1/ Boot board with both LVDS and HDMI (main) displays
    2/ Disconnect HDMI monitor and restart weston to work on LVDS as
main display
    3/ Re-connect HDMI display and restart weston to use both screens
again

- Error description:
    + Board couldn't get all modes from HDMI monitor
    + HDMI monitor could only display HD resolution while FHD is expected

Signed-off-by: Hung Dong <hung.dong.xd@rvc.renesas.com>
Signed-off-by: Nguyen Van Linh [FGA.BU13] <LinhNv25@fsoft.com.vn>
Signed-off-by: vietn <vietn@fsoft.com.vn>
Signed-off-by: Nguyen Van Tu [FGA.AIS] <TuNV16@fsoft.com.vn>
---
 drivers/gpu/drm/i2c/adv7511.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i2c/adv7511.c b/drivers/gpu/drm/i2c/adv7511.c
index bcac24a..404eb7d 100644
--- a/drivers/gpu/drm/i2c/adv7511.c
+++ b/drivers/gpu/drm/i2c/adv7511.c
@@ -577,6 +577,7 @@ static int adv7511_get_edid_block(void *data, u8 *buf, unsigned int block,
 	struct i2c_msg xfer[2];
 	uint8_t offset;
 	unsigned int i;
+	unsigned int cnt = 0;
 	int ret;
 
 	if (len > 128)
@@ -585,8 +586,17 @@ static int adv7511_get_edid_block(void *data, u8 *buf, unsigned int block,
 	if (adv7511->current_edid_segment != block / 2) {
 		unsigned int status, blk;
 
-		ret = regmap_read(adv7511->regmap, ADV7511_REG_DDC_STATUS,
-				  &status);
+		/* Use polling method to wait for DDC IDLE state.
+		 * This way we could avoid situations where
+		 * DDC controller is still stuck in Reading EDID state.
+		 */
+		do {
+			ret = regmap_read(adv7511->regmap, ADV7511_REG_DDC_STATUS,
+							&status);
+			msleep(10);
+			cnt++;
+		} while ((cnt < 200) && (status != 2));
+
 		if (ret < 0)
 			return ret;
 
-- 
2.7.4

